rfc-141-application-healthchecks.md

```diff
@@ -0,0 +1,224 @@
+probably fine.  The problem is elsewhere.
+
+Two of these indicate capacity problems outside of the instance, and
+the third is an alert that a manual maintenance procedure needs to be
+performed some time in the next two months.  They all share the
+property that if one instance reports a failure, all will.  This makes
+them unsuitable for load balancing purposes.
+
+Even if we don't make our healthchecks suitable for load balancing, as
+GOV.UK moves towards continuous deployments, we will end up in
+situations where an automatic deployment is aborted because of
+something unrelated to the change being deployed.  That's not good.
+
+A healthcheck should check that the instance is running and isn't
+completely broken.  That's all.
+
+## Proposal
+
+My proposal comes down to three parts:
+
+1. Remove the "warning" state of `GovukHealthcheck`, so a healthcheck
+   result is either "ok" or "critical".
+2. Make `GovukHealthcheck` serve an HTTP status of 500 for a
+   "critical" result.
+3. Add a separate healthcheck for liveness, to use with the ECS
+   container lifecycle.
+
+But we can't make this change right now, we have a migration path to
+follow first.
+
+### Remove the "warning" state
+
+There are four possible semantics we could assign the "warning" state:
+
+| # | State    | Allows automatic deployments? | Allows requests to be sent to the instance? |
+| - | -------- | ----------------------------- | ------------------------------------------- |
+|   | ok       | yes | yes |
+| 1 | warning  | yes | yes |
+| 2 | warning  | yes | no  |
+| 3 | warning  | no  | yes |
+| 4 | warning  | no  | no  |
+|   | critical | no  | no  |
+
+- If we go for option 1, then "warning" is the same as "ok".
+- If we go for option 2, then "warning" will let us deploy unusable releases.
+- If we go for option 3, then "warning" will block deployments.
+- If we go for option 4, then "warning" is the same as "critical".
+
+The most sensible option is 3, which is our current behaviour.
+
+But does it really gain us anything over having separate alerts for
+the specific condition we need to know about?
+
+By removing the "warning" state, we will remove some spurious alerts
+which don't add value, and add more specific alerts for ones which do.
+
+### Make `GovukHealthcheck` return a 500 status for critical failure
+
+This is a requirement for `GovukHealthcheck` to be usable with load
+balancers.
+
+### Have separate healthchecks for liveness and readiness
+
+If `GovukHealthcheck` checks whether an instance is ready to handle
+requests, and takes it out of the load balancer target group if not,
+then it is a *readiness* check.
+
+There is also value in having a separate *liveness* check, which
+simply checks if the instance is running and reachable, and doesn't
+check any of its dependencies:
+
+```ruby
+get "/healthcheck", to: proc { [200, {}, []] }
+get "/healthcheck/ready", to: GovukHealthcheck.rack_response(...)
```

Comment:
I feel like having two things called healthcheck enters some confusing waters. Particularly if one of the routes isn't named - it could be that the term "healthcheck" interchangeably could refer to either of these and it wouldn't be clear that `/healthcheck` refers to a liveness check rather than a ready check.

I wonder is there anything we can do to avoid 2 healthchecks? Are two definitely needed? If they are could we have one thing that is a healthcheck and one thing that is named something else? Then the term healthcheck doesn't become too convoluted (for instance with this change GovukHealthcheck no longer has anything to do with a liveness check)

If that doesn't work could we maybe use /healthcheck as a prefix and any checks within it have specific name (e.g `/healthcheck/app-running` and `/healthcheck/ready`). That could help by there being nothing that can be described as "the healthcheck" and that it's always a collective.

Reply:
That's a good point about the second healthcheck, @sengi what do you think?  Could we just have the readiness check (which returns either a 200 or a 500 status) and then the ECS healthchecks disregard the code and just check that the endpoint responds?

Reply:
It looks like the healthcheck is just a command: https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_HealthCheck.html

In which case, `[ "CMD-SHELL", "curl http://localhost/healthcheck" ]` will work.

I'll update the proposal to just have the one healthcheck.

Reply:
Since there's no longer an ECS-specific part to this proposal, I've added an appendix explaining why we want slightly different healthchecks for ECS, and how we can turn a readiness healthcheck into a liveness healthcheck.

Reply:
> Are two definitely needed?

Yes - separate liveness and readiness probes have been best practice for 5+ years now (basically since Kubernetes became popular - and they've been standard practice at Google for 10+ years). There are occasionally circumstances where you can get away with conflating them or omitting one or the other, but those are the exception rather than the rule - and we're talking about a general implementation across all GOV.UK apps here. Following best practices such as this will help to ensure we can run GOV.UK apps in containers reliably and without getting locked into to one particular system.

> I feel like having two things called healthcheck enters some confusing waters. Particularly if one of the routes isn't named - it could be that the term "healthcheck" interchangeably could refer to either of these and it wouldn't be clear that /healthcheck refers to a liveness check rather than a ready check.

I agree - /healthcheck would be a confusing choice of name for one of the healthchecks. I'd recommend deprecating the old endpoint and introducing new ones like you suggested. `/healthcheck/ready` is a good suggestion. At the risk of bikeshedding, I would probably go with `live` rather than `app-running` (so that it's clear that it's supposed to be used for the liveness probe). (If you look at examples of cloud-native open-source software you'll see liveness checks often called things like `/healthz` or `/health` but that's just a Google-ism that escaped into the outside world.)

On a practical note, I don't think ECS (or ALB, more precisely) is able to take 500 as a success response. Even if it could be made to do that, I think it would be surprising and confusing and lead to mistakes in future.

Reply:
In principle you could reuse the readiness check as a liveness check in ECS, but again I think that would be highly confusing and there would be no benefit.

Reply:
Ok, I've updated the proposal to define the two types of healthcheck, say we want to deprecate `/healthcheck` and set up `/healthcheck/live` + `/healthcheck/ready`, and give a motivating example of the sort of condition each check is supposed to detect.